### DISCUSSION TITLE
Removing polyfills from Naja

### DISCUSSION THREAD

**Discussion by jiripudil** [jiripudil/Naja#347] [General]
To provide out-of-the-box support for all browsers that were modern back then, Naja comes with two polyfills: for `AbortSignal` and for `EventTarget()` constructor. I figured now would be a good time to reevaluate if they are still necessary.

It turns out they are *not*.

## tl;dr

Naja will soon™ remove these polyfills from its dist package. They serve versions of browsers which – accumulated – account for not much more than 1 % of market share in 🇨🇿, according to caniuse data. The vast majority of users only download these bundled polyfills so that their browsers can throw them away.

**If you need to support legacy (non-Chromium) versions of Edge or some older versions of Safari (< 14) in your application, please include [the polyfills](https://github.com/naja-js/naja/blob/963c46a558e3574e0e26f47f7fc50902b1c9611e/src/polyfills.js) in your own application code.**

## What are they

`AbortSignal` adds support for abortable Fetch requests. Naja [supports this](https://github.com/naja-js/naja/blob/963c46a558e3574e0e26f47f7fc50902b1c9611e/src/Naja.ts#L111) and [even uses this itself](https://github.com/naja-js/naja/blob/963c46a558e3574e0e26f47f7fc50902b1c9611e/src/extensions/AbortExtension.ts#L29) in its built-in extensions. It has not been properly implemented in desktop Safari until 12.1, in iOS until 11.3, and in Edge until 16. According to [caniuse data](https://caniuse.com/abortcontroller) for 🇨🇿, these and older versions of the browsers amount to 0.14 % of the market in the case of iOS, and around 0.02–0.03 % in the case of both desktop Safari and Edge.

`EventTarget()` constructor is [crucial to the operation](https://github.com/naja-js/naja/blob/963c46a558e3574e0e26f47f7fc50902b1c9611e/src/Naja.ts#L17) of Naja's event system. It's a slightly more recent addition: it hasn't appeared in Edge until it switched to Chromium engine, and Apple implemented it in Safari/iOS 14. Again, according to [caniuse data](https://caniuse.com/mdn-api_eventtarget_eventtarget), this translates to about 0.1 % in the case of legacy Edge versions altogether, and 0.37 % and 0.64 % for Safari (desktop and iOS, respectively).

## Costs and benefits

Accumulated, the affected browser versions account for approximately 1.1 % of the market share. I believe the number is likely to decrease further: legacy Edge has been deprecated and Microsoft [actively pushes](https://www.zdnet.com/article/windows-10-microsoft-begins-automatically-pushing-chromium-edge-to-users/) the new Chromium-based Edge towards users, and Apple has released Safari/iOS/iPadOS 14 for systems and devices that are nearing their end-of-life (2015 models of iPhone and iPad, and 2012 Macs).

On the other side of the equation, the benefits appear to be non-negligible: removing the polyfills cuts Naja's dist size in half, from ~35 kB down to ~17 kB minified and from ~10 kB to ~5 kB minzipped. Considering the aforementioned marginal market share of the targeted browser versions, this means that up to 99 % of users unnecessarily download twice as much data which is then unused anyway.

From where I see it, all available data speak unanimously in favor of removing the polyfills.

## Towards a modern web platform

It has been the goal of Naja 2.x to build upon common, well-known, standardized web APIs such as `fetch()`, `AbortSignal`, `EventTarget`, and `CustomEvent`s. Naja has included the polyfills in question so that you could grab its minified dist file, plop it into your website, and it would work, *everywhere*.

It's been a year and a half since the release of 2.0.0 and the adoption of those APIs has shifted. In my eyes, 99 % is a pretty good definition of 'everywhere.' I think it's fair for Naja to expect to run in a modern environment where all standard APIs are implemented and made available by the browser, especially when a huge majority of visits, statistically speaking, are made using a capable browser.

## What to do

Go take a look into your analytics dashboard and figure out which browsers your visitors use to access your website. If they do not include a portion of legacy Edge or Safari < 14 significant enough by your standards, you can safely upgrade Naja once it ships without the polyfills, and be happy that you're serving your users a smaller client-side bundle.

Should you need to support the problematic browser versions, you can very easily [include the polyfills](https://github.com/naja-js/naja/blob/963c46a558e3574e0e26f47f7fc50902b1c9611e/src/polyfills.js) in your own application codebase and Naja will be able to pick them up and use them if necessary.

## The timeline

I'm going to leave this announcement here for two weeks and unless it gathers a significant amount of criticism and counter-arguments during the period, I'm going to release a polyfill-free Naja 2.3.0 on December 23, as an early Christmas present.

**Comment by dakur** [comment]
Why not 3.0.0? If someone `yarn upgrade` and doesn't know about this change, it stop work at once for him.

**Reply by jiripudil** [reply]
Hi, that's something I was considering too, and carefully. This removal of polyfills doesn't introduce any change in the API, and Naja's compatibility promise, as declared in the readme and the docs, remains fulfilled:

> [Naja is] compiled into a bundle that works in all modern browsers. Every build of Naja is thoroughly tested and therefore guaranteed to work in the latest versions of Chromium (Chrome and Edge), Firefox, and WebKit (Safari).

As to browsers/systems that this next release will render incompatible without polyfills:

- I've already argued why Naja does not consider legacy Edge a modern browser in #99.
- Safari 14 is available since macOS Mojave which, as I've mentioned above, supports Mac models that go back to 2012. The preceding version, macOS High Sierra, has received its last security update in November 2020, and can be considered officially unsupported since.
- iOS/iPadOS 14 that include Safari 14 are available for *all* devices that support the previous version 13, and are therefore not receiving any more security updates either.

I believe dropping support for old browsers which are not even any longer supported by their vendors is reasonably safe to do in a minor version, especially when their newer versions are so widely available.
